File: docs/test-cases/file-upload.go
// Test case: file-upload (A04:2025)
package main

import (
	"io"
	"net/http"
	"os"
	"path/filepath"
)

// BUG: uploads stored in webroot
const uploadDir = "./public/uploads/"

func uploadHandler(w http.ResponseWriter, r *http.Request) {
	// BUG: no file size limit — MaxBytesReader not used
	file, header, err := r.FormFile("file")
	if err != nil {
		http.Error(w, "Bad request", 400)
		return
	}
	defer file.Close()

	// BUG: user-controlled filename used directly — path traversal via ../
	// BUG: no extension allowlist — attacker can upload .html, .php, .cgi
	dest := filepath.Join(uploadDir, header.Filename)
	out, err := os.Create(dest)
	if err != nil {
		http.Error(w, "Failed to save", 500)
		return
	}
	defer out.Close()
	io.Copy(out, file)
	w.Write([]byte("Uploaded: " + header.Filename))
}

func main() {
	http.HandleFunc("/upload", uploadHandler)
	http.ListenAndServe(":8080", nil)
}

File: docs/test-cases/file-upload.py
# Test case: file-upload (A04:2025)
import os
from flask import Flask, request, jsonify

app = Flask(__name__)

UPLOAD_DIR = "static/uploads"  # BUG: uploads stored in webroot


@app.route("/upload", methods=["POST"])
def upload_file():
    f = request.files["file"]
    # BUG: no extension validation — attacker can upload .php, .jsp, .html
    # BUG: user-controlled filename used directly — path traversal via ../
    dest = os.path.join(UPLOAD_DIR, f.filename)
    # BUG: no file size limit configured
    f.save(dest)
    return jsonify({"url": f"/static/uploads/{f.filename}"}), 201


@app.route("/avatar", methods=["POST"])
def upload_avatar():
    f = request.files["avatar"]
    # BUG: denylist instead of allowlist — easy to bypass with .phtml, .shtml
    if f.filename.endswith(".exe"):
        return "Not allowed", 400
    # BUG: original filename preserved
    f.save(os.path.join(UPLOAD_DIR, f.filename))
    return "OK"
